File: site/scripts/application-package.mjs
import { readFile, lstat, readdir } from 'node:fs/promises';
import { join, resolve } from 'node:path';
import { createHash } from 'node:crypto';

export const sha256 = bytes => createHash('sha256').update(bytes).digest('hex');
const digestPattern = /^[a-f0-9]{64}$/;
const profileKeys = ['siteUrl', 'base', 'audience', 'editApi', 'canvasBackend'];

/** digest/profile come from trusted installation or CI provenance, never Canvas or this directory. */
export async function verifyApplicationPackage(directory, { digest, profile } = {}) {
  if (!digestPattern.test(digest ?? '')) throw new Error('Expected approved package digest is required');
  const root = resolve(directory);
  if (!(await lstat(root)).isDirectory() || !(await lstat(join(root, 'package.json'))).isFile()) throw new Error('Non-regular application package');
  const bytes = await readFile(join(root, 'package.json'));
  if (sha256(bytes) !== digest) throw new Error('Application manifest digest mismatch');
  const manifest = JSON.parse(bytes);
  if (manifest.type !== 'module' || manifest.protocol !== 1 || !/^[a-f0-9]{40}$/.test(manifest.source) || !digestPattern.test(manifest.dependencyDigest ?? '')) throw new Error('Incompatible application protocol');
  if (manifest.nodeMajor !== Number(process.versions.node.split('.')[0])) throw new Error('Incompatible application runtime');
  if (!manifest.profile || profileKeys.some(key => typeof manifest.profile[key] !== 'string') ||
      !['internal', 'public'].includes(manifest.audience) || manifest.profile.audience !== manifest.audience || manifest.profile.base !== manifest.base ||
      !/^\/(?:[A-Za-z0-9_-]+\/)*$/.test(manifest.base) ||
      (profile && profileKeys.some(key => profile[key] !== manifest.profile[key]))) throw new Error('Application profile mismatch');
  if (!Array.isArray(manifest.files) || manifest.files.length > 20000) throw new Error('Invalid application inventory');
  const seen = new Set(), normalized = new Set();
  let total = 0;
  const checks = [];
  for (const file of manifest.files) {
    if (typeof file.path !== 'string' || !/^(?:private|public)\//.test(file.path) ||
        file.path.includes('\\') || /[\u0000-\u001f]/.test(file.path) || file.path.split('/').some(p => !p || p === '.' || p === '..') ||
        normalized.has(file.path.toLowerCase())) throw new Error('Invalid application path or collision');
    if (!Number.isSafeInteger(file.size) || file.size < 0 || !digestPattern.test(file.hash ?? '')) throw new Error('Invalid application inventory');
    total += file.size;
    if (total > 512 * 1024 * 1024) throw new Error('Application package exceeds size budget');
    seen.add(file.path); normalized.add(file.path.toLowerCase());
    checks.push(async () => {
      let path = root;
      const parts = file.path.split('/');
      for (let index = 0; index < parts.length; index++) {
        path = join(path, parts[index]);
        const stat = await lstat(path);
        if (index < parts.length - 1 ? !stat.isDirectory() : !stat.isFile()) throw new Error('Non-regular or symbolic application path');
      }
      const content = await readFile(path);
      if (content.length !== file.size || sha256(content) !== file.hash) throw new Error('Application checksum mismatch');
    });
  }
  // Bound concurrent reads while retaining every path, byte and inventory check.
  for (let offset = 0; offset < checks.length; offset += 4) {
    const results = await Promise.allSettled(checks.slice(offset, offset + 4).map(check => check()));
    const failure = results.find(result => result.status === 'rejected');
    if (failure) throw failure.reason;
  }
  for (const required of ['private/renderer.mjs', 'private/template.html', 'private/template-docs.html', 'public/.vite/manifest.json'])
    if (!seen.has(required)) throw new Error('Incomplete application inventory');
  async function checkInventory(relative = '') {
    for (const entry of await readdir(join(root, relative), { withFileTypes: true })) {
      const path = relative ? `${relative}/${entry.name}` : entry.name;
      if (entry.isDirectory()) await checkInventory(path);
      else if (!entry.isFile() || (path !== 'package.json' && !seen.has(path))) throw new Error('Application inventory contains unlisted or non-regular files');
    }
  }
  await checkInventory();
  return manifest;
}

File: edit-service/application_package.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

var packageDigestPattern = regexp.MustCompile(`^[a-f0-9]{64}$`)

// Only explicit, authenticated CI promotion may write this pointer. Canvas
// descriptors select a release; they are never authority to install executable
// code. Install packages immutably and keep the pointer read-only to the service.
type approvedApplication struct {
	Directory     string `json:"directory"`
	Digest        string `json:"digest"`
	Source        string `json:"source"`
	Repo          string `json:"repo"`
	WorkflowRunID int64  `json:"workflowRunId"`
}

func readRegularBounded(path string, limit int64) ([]byte, error) {
	info, err := os.Lstat(path)
	if err != nil {
		return nil, err
	}
	if !info.Mode().IsRegular() || info.Size() > limit {
		return nil, errors.New("invalid package file")
	}
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	b, err := io.ReadAll(io.LimitReader(f, limit+1))
	if int64(len(b)) > limit {
		return nil, errors.New("package file too large")
	}
	return b, err
}

func loadApprovedApplication(c localBuildConfig, repo string) (approvedApplication, error) {
	var app approvedApplication
	if !filepath.IsAbs(c.ApplicationPointer) {
		return app, errors.New("absolute application pointer required")
	}
	b, err := readRegularBounded(c.ApplicationPointer, 16*1024)
	if err != nil {
		return app, err
	}
	if err = json.Unmarshal(b, &app); err != nil {
		return app, err
	}
	if !filepath.IsAbs(app.Directory) || !packageDigestPattern.MatchString(app.Digest) || !gitSHA.MatchString(app.Source) || app.Repo != repo || app.WorkflowRunID <= 0 {
		return app, errors.New("invalid approved application provenance")
	}
	info, err := os.Lstat(app.Directory)
	if err != nil || !info.IsDir() {
		return app, errors.New("invalid application directory")
	}
	b, err = readRegularBounded(filepath.Join(app.Directory, "package.json"), 8*1024*1024)
	if err != nil {
		return app, err
	}
	hash := sha256.Sum256(b)
	if hex.EncodeToString(hash[:]) != app.Digest {
		return app, errors.New("application manifest checksum mismatch")
	}
	var manifest struct {
		Type             string       `json:"type"`
		Protocol         int          `json:"protocol"`
		Source           string       `json:"source"`
		NodeMajor        int          `json:"nodeMajor"`
		Base             string       `json:"base"`
		Audience         string       `json:"audience"`
		DependencyDigest string       `json:"dependencyDigest"`
		Profile          buildProfile `json:"profile"`
		Files            []struct {
			Path string `json:"path"`
			Hash string `json:"hash"`
			Size int64  `json:"size"`
		} `json:"files"`
	}
	if err = json.Unmarshal(b, &manifest); err != nil {
		return app, err
	}
	if manifest.Type != "module" || manifest.Protocol != 1 || manifest.Source != app.Source || manifest.NodeMajor != 24 || manifest.Profile != c.Profile || manifest.Base != c.Profile.Base || manifest.Audience != c.Profile.Audience || !packageDigestPattern.MatchString(manifest.DependencyDigest) || len(manifest.Files) > 20000 {
		return app, errors.New("incompatible application package")
	}
	seen, folded := map[string]bool{}, map[string]bool{}
	var total int64
	for _, file := range manifest.Files {
		if (!strings.HasPrefix(file.Path, "private/") && !strings.HasPrefix(file.Path, "public/")) || strings.ContainsAny(file.Path, "\\\x00\r\n") || folded[strings.ToLower(file.Path)] || file.Size < 0 || file.Size > 512*1024*1024 || !packageDigestPattern.MatchString(file.Hash) {
			return app, errors.New("invalid application inventory")
		}
		total += file.Size
		if total > 512*1024*1024 {
			return app, errors.New("application size budget exceeded")
		}
		path := app.Directory
		parts := strings.Split(file.Path, "/")
		for i, part := range parts {
			if part == "" || part == "." || part == ".." || strings.ContainsFunc(part, func(r rune) bool { return r < 32 }) {
				return app, errors.New("invalid application path")
			}
			path = filepath.Join(path, part)
			info, err := os.Lstat(path)
			if err != nil || (i < len(parts)-1 && !info.IsDir()) || (i == len(parts)-1 && !info.Mode().IsRegular()) {
				return app, errors.New("non-regular application path")
			}
		}
		// Hash through a bounded stream: verification does not allocate the whole
		// package in the editor's memory budget.
		f, err := os.Open(path)
		if err != nil {
			return app, err
		}
		h := sha256.New()
		n, copyErr := io.Copy(h, io.LimitReader(f, file.Size+1))
		closeErr := f.Close()
		if copyErr != nil || closeErr != nil || n != file.Size || hex.EncodeToString(h.Sum(nil)) != file.Hash {
			return app, errors.New("application file checksum mismatch")
		}
		seen[file.Path], folded[strings.ToLower(file.Path)] = true, true
	}
	for _, required := range []string{"private/renderer.mjs", "private/template.html", "private/template-docs.html", "public/.vite/manifest.json", "private/commands/prepare-content.mjs", "private/commands/coordinate.mjs", "private/commands/staged.mjs", "private/commands/application-package.mjs", "private/commands/client-assets.mjs", "private/commands/content-output.mjs", "private/commands/build-item-history.mjs", "private/commands/check-demo.mjs", "private/commands/check-document-links.mjs", "private/commands/check-item-history.mjs", "private/commands/validate_items.py"} {
		if !seen[required] {
			return app, errors.New("incomplete application package")
		}
	}
	err = filepath.WalkDir(app.Directory, func(path string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if entry.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(app.Directory, path)
		if err != nil {
			return err
		}
		if !entry.Type().IsRegular() || (rel != "package.json" && !seen[filepath.ToSlash(rel)]) {
			return errors.New("unlisted application file")
		}
		return nil
	})
	return app, err
}
